Added spike archiving and support for stdp-synapses to precise neurons

heplesser opened the pull request:
This should fix #85.

- Spike times are archived as actual spike time, i.e., stamp at end of interval minus offset.
- This commit implies on semantic change: for precise neurons, `t_spike` now is the actual spike time, not the end of the time step
- The `offset` is no longer returned by `GetStatus` on precise models.
- Since both `t_spike` and `offset` are undocumented properties, this should not be much of a problem.
- Two tests adapted and one regression test added.

Suggested reviewers: @suku248 @abigailm I'd appreciate if you could review this one quickly, since my fix for #77 depends on it (in particular on parrot_neuron_ps).

tammoippen commented on precise/parrot_neuron_ps.h:
```diff
@@ -34,38 +34,32 @@ Name: parrot_neuron_ps - Neuron that repeats incoming spikes handling
+through port 1 will be ignored. This allows setting exact pre-
+and post-synaptic spike times for STDP protocols by connecting
+two parrot neurons spiking at desired times by, e.g., a
+stdp_synapse onto port 1 on the post-synaptic parrot neuron.
```
I believe, this is not true, yet. You changed `parrot_neuron_ps::handles_test_event(...)` to accept connections to port 1, but `parrot_neuron_ps::handle(...)` is unchanged, adding all incoming events, not differentiating between ports.

tammoippen commented:
Nice work! I have not looked at the sli code in detail, though. You get my 'thumbs up', when the points with the `parrot_neuron_ps` are addressed.

heplesser commented:
@tammoippen Good catch! I now added a proper test and fixed some more issues. `parrot_neuron_ps` now passes the same set of tests as `parrot_neuron` wrt handling rports, multiplicity and plasticity of synpases between parrots.

heplesser commented:
To test that spike multiplicity is handled properly in conjunction with STDP, I have now added a test that compares weight changes in an stdp_synapse between two parrot neurons and two precise parrot neurons, when multiple spikes occur during a single time step. This is based on an idea in #77, but is included in the PR for this issue, since it depends on rport 1 support in parrot_neuron_ps. The logic of the test is described in detail in the docstring of the test class.

Thanks, @flinz, for coming up with the idea of placing plastic synapses between parrot neurons for testing!

flinz commented:
@heplesser sure, I'm happy its of use!

tammoippen commented:
:+1: from my side!

krishnanj commented on precise/iaf_psc_exp_ps.cpp:
```diff
@@ -458,6 +452,7 @@ nest::iaf_psc_exp_ps::emit_spike_( const Time& origin,
+  set_spiketime( Time::step( S_.last_spike_step_ ), S_.last_spike_offset_ );
   SpikeEvent se;
 
   se.set_offset( spike_offset );
```
se.set_offset( S_.last_spike_offset_ );

krishnanj replied:
Only a minor change I think, otherwise :thumbsup:

heplesser commented:
@krishnanj Thanks for your review, I fixed this now by completely eliminating the local variable.

heplesser commented:
This PR should be merged after #160.

krishnanj commented:
:+1:

abigailm commented:
:+1:  will merge once #160 is merged

heplesser commented:
@abigailm Thanks for your review! I see that this PR now has conflicts, possibly due to the merge of #160, or #171. I will explore this later today, got a plane to catch now.

tammoippen commented:
Very nice work! Merging.
